### PR TITLE
[Structural] Adding Plastic-Damage Model & minor error solvings

### PR DESCRIPTION
**Pull request by SergioJimenezReyes**
Plastic-damage CL used in PLCd and presented in Lucioni, B. et al (1995).
While doing this CL @AlejandroCornejo has found an error which affects the isotropic plasticity CL and the kinematic plasticity CL and is corrected in this PR.

### CODE REVIEW

**Comment by loumalouomega** [comment]
@AlejandroCornejo check Travis, is failing

**Comment by AlejandroCornejo** [comment]
> @AlejandroCornejo check Travis, is failing

YES! I'm changing the tests values and so on. Working on it

**Comment by loumalouomega** [comment]
> > @AlejandroCornejo check Travis, is failing
> 
> YES! I'm changing the tests values and so on. Working on it

Check that actually works, and check the validation tests too

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h [hunk 1]
```diff
@@ -815,7 +815,7 @@ class GenericConstitutiveLawIntegratorKinematicPlasticity
         double& rHardeningParameter
         )
     {
-        rHardeningParameter = -SlopeThreshold;
+        rHardeningParameter = SlopeThreshold;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:818]: Ups

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp [hunk 2]
```diff
@@ -0,0 +1,1000 @@
...
+                damage_dissipation, r_constitutive_matrix,
+                rValues, characteristic_length, damage_yield_flux,
+                plastic_strain, damage, damage_increment, 
+                undamaged_free_energy, hardd, damage_dissipation_increment);
+
+        // Verification threshold for the plastic-damage process
+        if (plasticity_indicator >= std::abs(1.0e-4 * threshold_plasticity) && damage_indicator >= std::abs(1.0e-4 * threshold_damage)) {
+            bool is_converged = false;
+            int number_iteration = 0;
+            const int max_iter = 100;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp:131]: Can we make this not hardcoded?

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h (hunk 1 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:818]: image

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp [hunk 3]
```diff
@@ -0,0 +1,1000 @@
...
+    if ((check_base + check_integrator_plasticity + check_integrator_damage) > 0) return 1;
+    return 0;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+template <class TPlasticityIntegratorType, class TDamageIntegratorType>
+double GenericSmallStrainPlasticDamageModel<TPlasticityIntegratorType, TDamageIntegratorType>::
+CalculateDamageParameters(
+	array_1d<double, 6>& rPredictiveStressVector,
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp]: Please, can you create structures?, this quantity of arguments is too much

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp [hunk 4]
```diff
@@ -0,0 +1,1000 @@
...
+/***********************************************************************************/
+template <class TPlasticityIntegratorType, class TDamageIntegratorType>
+double GenericSmallStrainPlasticDamageModel<TPlasticityIntegratorType, TDamageIntegratorType>::
+CalculatePlasticParameters(
+    array_1d<double, VoigtSize>& rPredictiveStressVector,
+    Vector& rStrainVector,
+    double& rUniaxialStress,
+    double& rThreshold,
+    double& rPlasticDenominator,
+    array_1d<double, VoigtSize>& rFflux,
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp]: Idem, too many values

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.h [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+     * @param CharacteristicLength The equivalent length of the FE
+     * @param rPlasticStrain The plastic component of the strain
+     * @param Damage The internal variable of the damage model
+     */
+    double CalculatePlasticParameters(
+        array_1d<double, VoigtSize>& rPredictiveStressVector,
+        Vector& rStrainVector,
+        double& rUniaxialStress,
+        double& rThreshold,
+        double& rPlasticDenominator,
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.h]: aaaaah, save me

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.h [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
+    // Converged values
+    double mPlasticDissipation = 0.0;
+    double mThresholdPlasticity = 0.0;
+    Vector mPlasticStrain = ZeroVector(VoigtSize);
+	double mThresholdDamage = 0.0;
+	double mDamage = 0.0;
+	double mDamageDissipation = 0.0;
+
+    // only for printing
+    double mUniaxialStress = 0.0;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.h:556]: Please remove if only is post post process

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_kinematic_plasticity_constitutive_laws_3d.cpp [hunk 7]
```diff
@@ -112,6 +112,8 @@ KRATOS_TEST_CASE_IN_SUITE(ConstitutiveLawIntegrateStressPlasticitySmallStrainKin
     TestVM = cl_parameters.GetStressVector();
     VonMisesCL.GetValue(PLASTIC_DISSIPATION, plastic_dissipation);
     KRATOS_WARNING_IF("TestPlasticity", plastic_dissipation < 1.0e-12) << "VonMises:: This test is not in plastic range" << std::endl;
+	
+	KRATOS_WATCH(TestVM)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_kinematic_plasticity_constitutive_laws_3d.cpp]:
```suggestion
```

Clean this

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Look at my comments

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp (hunk 2 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp:131]: it converges in 2 iterations usually... How can we modify this?

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp]: ok

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp:131]: With a property, but keep it as it is now, and we fix it for the others later

**Comment by loumalouomega** [comment]
Travis still fails

**Comment by loumalouomega** [comment]
Please, check the validation test https://github.com/KratosMultiphysics/Kratos/blob/e9dfdaa19c7fc116447ae86c2b13c58ceec69351/applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py#L426

**Comment by AlejandroCornejo** [comment]
> Please, check the validation test
> 
> [Kratos/applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py](https://github.com/KratosMultiphysics/Kratos/blob/e9dfdaa19c7fc116447ae86c2b13c58ceec69351/applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py#L426)
> 
> Line 426 in [e9dfdaa](/KratosMultiphysics/Kratos/commit/e9dfdaa19c7fc116447ae86c2b13c58ceec69351)
> 
>  validationSuite.addTest(TTensileTestStructuralTest('test_execution'))

done

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.h (hunk 5 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.h]: solved

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp (hunk 4 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp]: solved

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_plastic_damage_model.cpp]: solved now

**Review by loumalouomega — APPROVED** [review]

**Comment by loumalouomega** [comment]
There is a tiny complain from Codacy

https://app.codacy.com/app/msandre/Kratos/pullRequest?prid=3421858

**Pull request merged**